webob/response.py

```diff
@@ -225,43 +222,66 @@ def __init__(self, body=None, status=None, headerlist=None, app_iter=None,
         # Response with empty body, such as Response(status='204 No Content')
         # without the default content_type being set
 
-        if (
-            self.content_type is None and
-            headerlist is None and
-            _code_has_body(self.status_code)
-        ):
-            self.content_type = content_type
+        encoding = None
+        code_has_body = (
+            self._status[0] != '1' and
+            self._status[:3] not in ('204', '205', '304')
+        )
+        if headerlist is None and code_has_body:
+            content_type = content_type or self.default_content_type
 
-        # Set up the charset
-        #
-        # In contrast with the above, if a charset is not set but there is a
-        # content_type we will set the default charset if the content_type
-        # allows for a charset.
+            # Check if content_type is set because default_content_type could
+            # be None, in which case there is no content_type
+            if content_type:
+
+                # Set up the charset, if the content_type doesn't already have one
 
-        if self.content_type:
-            if not self.charset and charset is not _marker:
-                self.charset = charset
-            elif not self.charset and self.default_charset:
-                if _content_type_has_charset(self.content_type):
-                    self.charset = self.default_charset
+                has_charset = 'charset=' in content_type
+
+                if not has_charset and charset is not _marker and charset is not None:
+                    encoding = charset
+                    content_type += '; charset=' + charset
+                elif not has_charset and charset is _marker and self.default_charset:
+                    # Optimize for the default_content_type case, avoid a function call
+                    if (
+                        content_type is self.default_content_type or
+                        _content_type_has_charset(content_type)
+                    ):
+                        encoding = self.default_charset
+                        content_type += '; charset=' + self.default_charset
+
+            self._headerlist.append(('Content-Type', content_type))
 
         # Set up conditional response
         if conditional_response is None:
             self.conditional_response = self.default_conditional_response
         else:
             self.conditional_response = bool(conditional_response)
 
-        # Set up app_iter
-        if app_iter is None:
+        # Set up app_iter if the HTTP Status code has a body
+        if app_iter is None and code_has_body:
             if isinstance(body, text_type):
-                encoding = self.charset
+                # Fall back to trying self.charset if encoding is not set. In
+                # most cases encoding will be set to the default value.
+                encoding = encoding or self.charset
```

Comment:
I don't think so, I would love to hear arguments for/against it... but `default_body_encoding` is strictly for `.text` to me. 

`__init__` won't let you create a `204 No Content` with a body for example (because setting Content-Length on such a response is not allowed by the HTTP spec), `__init__` will do it's best to give you a valid `Response`. If you make changes/set invalid options on the `Response` after the fact, then so be it.

There is no way to influence `default_body_encoding` in `__init__` so you end up using the default all of the time. I'd recommend encoding before it reaches `Response`. Changing this would also be b/w incompatible, if you set the `content_type` to `image/jpg` in older versions of WebOb you couldn't provide a text body because there would be no charset set.

Reading through the 1.6.1 code though, it seems if you set `charset` and `content_type` it would happily create `image/jpg; charset=utf-8` and then use the charset if passed in. So I guess I could use the provided `charset` as if it were the `default_body_encoding`, even-though it is not added the `content_type`.

If you do `Response(text='sometext')` then `__init__` will call the `.text` attribute and `default_body_encoding` will be used to set the `body` up though...